packages/context/src/binding.ts

```diff
@@ -92,6 +96,15 @@ export enum BindingType {
   PROVIDER = 'Provider',
 }
 
+export type BindingOptions =
+  | {[name: string]: BoundValue}
+  | Promise<{[name: string]: BoundValue}>;
```

Comment:
I am little bit worried about having a deferred options object as a whole. It means sync functions like `findByTag` cannot inspect `BindingOptions`, which defeats the purpose of generic options IMO.

Is it really necessary to accept deferred options?

If it is, can we move the deferred part into individual option values instead?

```ts
interface BindingOptions {
  {[name: string]: ValueOrPromise<BoundValue>;
}
```

This way we can add predefined options to the interface (once we come with some), with a finer-grained control of what can be sync and what can be supplied via a Promise. For example, we can move `scope` and `tags` to `options`:

```ts
interface BindingOptions {
  tags: Set<string>;
  scope: BindingScope;
  {[name: string]: ValueOrPromise<BoundValue>;
}
```

Reply:
Haha, this comment contradicts my earlier https://github.com/strongloop/loopback-next/pull/671#issuecomment-338691103

> The implementation is too complex to my taste, a big part of this complexity is caused by code dealing with promises and nested promises in options properties. IMO, we should not resolve promises at options-property level. Users wishing to build options from individual promises should use a helper function to convert an object containing promise properties into a promise of an object with resolved properties.

I guess your current proposal is the way to go then!

I still think the type definition can and should be simplified a bit:

```ts
export type BindingOptions = ValueOrPromise<{
  [name: string]: ValueOrPromise<BoundValue>;
}>
```